[WIP]Initial GPU acceleration support for LightGBM

huanzhang12 opened the pull request:
This set of patches adds initial GPU acceleration for LightGBM, by accelerating histogram construction on GPUs. The implementation is highly modular and does not affect existing features of LightGBM; GPU kernel code is mostly independent of other parts of LightGBM, so long-term maintenance is easier.

I add a new type of tree learner in gpu_tree_learner.h and gpu_tree_learner.cpp. All GPU interfacing code is inside these two files. Changes to other parts of LightGBM are kept minimal; I only make some small interface changes to make necessary data structure available for the GPU tree learner. GPU code (in folder src/treelearner/ocl/) are implemented in OpenCL and tested on both AMD and NVIDIA GPUs.

For building and testing procedure, and some initial performance comparison, please see:
https://github.com/huanzhang12/lightgbm-gpu
(We probably want to move these instructions to Wiki after merging)

msftclas commented:
@__huanzhang12__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@huanzhang12__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

chivee commented:
Hi @huanzhang12 , thanks for your great contributions. could you please also share some experiments result to make it clear to the users of lightGBM

chivee commented:
|                   | CPU 255 bins | CPU 63 bins | CPU 15 bins | GPU 255 bins | GPU 63 bins | GPU 15 bins |
|-------------------|--------------|-------------|-------------|--------------|-------------|-------------|
| Higgs AUC         | 0.846132     | 0.845311    | 0.84087     | 0.84523      | 0.845428    | 0.840548    |
| Epsilon AUC       | 0.950452     | 0.950097    | 0.94823     | 0.950199     | 0.950109    | 0.947978    |
| Expo AUC          | 0.776091     | 0.770014    | 0.743602    | 0.775943     | 0.772696    | 0.743317    |
| MS-LTR NDCG@1     | 0.521135     | 0.523852    | 0.517996    | 0.521614     | 0.521099    | 0.52055     |
| MS-LTR NDCG@3     | 0.504635     | 0.505756    | 0.502495    | 0.505045     | 0.50397     | 0.503641    |
| MS-LTR NDCG@5     | 0.508776     | 0.510531    | 0.507656    | 0.509656     | 0.50904     | 0.50889     |
| MS-LTR NDCG@10    | 0.526651     | 0.52749     | 0.524527    | 0.527411     | 0.526226    | 0.525855    |
| Yahoo-LTR NGCG@1  | 0.732531     | 0.729834    | 0.732634    | 0.73133      | 0.732279    | 0.731965    |
| Yahoo-LTR NGCG@3  | 0.739642     | 0.737349    | 0.737533    | 0.737912     | 0.738382    | 0.736739    |
| Yahoo-LTR NGCG@5  | 0.756766     | 0.755554    | 0.755182    | 0.755944     | 0.756303    | 0.754295    |
| Yahoo-LTR NGCG@10 | 0.796924     | 0.795605    | 0.795454    | 0.796339     | 0.796369    | 0.795321    |


![Performance Comparison](http://www.huan-zhang.com/images/upload/lightgbm-gpu/compare_9602cd7e.png)

huanzhang12 commented:
@chivee Currently all results are posted in my development repository: https://github.com/huanzhang12/lightgbm-gpu
(See the "Performance Comparison" section)

Do you want to to copy the results here? After my changes are reviewed and merged, I think we can then move these instructions and results to the Wiki page.

chivee commented:
@huanzhang12 , I have copied them to here for a better review. Yes, these experiments result could be a huge guideline for the users.

guolinke commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "include/boost/compute"]
+	path = compute
+	url = https://github.com/boostorg/compute
```
can we avoid to use boost? it seems you only use it for the aligned_alloc .

huanzhang12 replied:
But if GPU support is not enabled at compile time, Boost is not required.

Laurae2 replied:
@guolinke Can you check for the licensing?

Boost licensing shouldn't be an issue as Boost users are free to do whatever they do with it. The only obligation is to not remove the Boost license from the submodule (if you create compiled code afterwards, then putting the Boost license is not required - it must be only there before compiling code).

Users are able to fetch the LightGBM repository without Boost using `recursive`, while those who want the full repository can use `recursive` on git.

huanzhang12 commented:
@chivee Thanks!

guolinke commented on include/LightGBM/feature_group.h:
```diff
@@ -181,6 +185,8 @@ class FeatureGroup {
   /*! \brief True if this feature is sparse */
   bool is_sparse_;
+  /*! \brief Threshold for treating a feature as a sparse feature */
+  double sparse_threshold_;
```
It seems the ```sparse_threshold_``` only use in the constructor. I think we don't need to save it.

huanzhang12 replied:
Fixed in 12e5b82

guolinke replied:
I think it is better to remove this member variable(sparse_threshold_) as well.

huanzhang12 replied:
Currently this member variable is used in Dataset::CreateValid() and Dataset::CopyFeatureMapperFrom(). When copying a Dataset object to a new Dataset object we need to know the sparse_threshold of the old Dataset to reconstruct the FeatureGroup objects.

guolinke replied:
@huanzhang12 Sorry, I didn't see it.
Ok,  I think we can put it into ```Dataset``` class to avoid the duplicate.

huanzhang12 replied:
The reason I put `sparse_threshold` into `FeatureGroup` is that a very related parameter, `is_enable_sparse`, is also a member per `FeatureGroup` rather than per `Dataset`. When I was adding `sparse_threshold` I wanted to manage the two parameters in the same way. 

I can certainly put sparse_threshold into `Dataset` instead if you request me to do so. But is there a reason why `is_enable_sparse` is not put into `Dataset`? Or maybe its just an imperfection of the code?

guolinke replied:
I think the ```is_enable_sparse``` you said is the ```is_sparse``` ? Not all ```feature_group``` are sparse, so put it into feature_group is reasonable. But all feature_group share the same sparse_threshold. 
As for the ```is_enable_sparse```, it is a member in the config class, not be saved in the feature_group.

huanzhang12 replied:
Thanks for the clarification! I have moved `sparse_threshold` to the `Dataset` class in commit `cb35a028`

guolinke commented on include/LightGBM/config.h:
```diff
@@ -188,6 +189,9 @@ struct TreeConfig: public ConfigBase {
   int top_k = 20;
+  int gpu_platform_id = -1;
+  int gpu_device_id = -1;
+  bool gpu_use_dp = false;
```
Add some details for these settings ?

huanzhang12 replied:
Fixed in 1159854

huanzhang12 commented:
The Travis CI build failure is on a pep8 problem inside boost.compute (now a sub-module).
I think we want to exclude the "compute" directory when doing the pep8 check?

guolinke commented:
@wxchan can you help to look the pep8 check problem?

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1207 @@
+        host4[j].s2 = (iters[4].RawGet(j) * dev_bin_mult[4] + ((j+4) & (dev_bin_mult[4] - 1))) 
+                    |((iters[5].RawGet(j) * dev_bin_mult[5] + ((j+5) & (dev_bin_mult[5] - 1))) << 4);
+        host4[j].s3 = (iters[6].RawGet(j) * dev_bin_mult[6] + ((j+6) & (dev_bin_mult[6] - 1))) 
+                    |((iters[7].RawGet(j) * dev_bin_mult[7] + ((j+7) & (dev_bin_mult[7] - 1))) << 4);
```
It seems you call ```RawGet``` many times. 
I think a better solution is use a function like ```CopyTo(buffer)``` to get these bin data by only one function call.

huanzhang12 replied:
This part of code makes 4 (or 8) features to a 32-bit feature tuple. So 4 (or 8) features are bundled to one here; a 32-bit host4[i] consists of 4 (or 8) features. CopyTo() seems difficult here, as the destination memory is not linear. We can probably copy each feature data to a temporary array, and use these arrays to build the feature bundles, but this actually has more overhead if these RawGet() can be inlined.

Performance-wise, these RawGet()'s should be inlined by compiler (I checked gcc generated assembly and it is very efficient, but not quite sure on Windows), and it is equivalent to accessing the data_ array of BinIterator using the given index. The tricky part is to make the RawGet() function call non-virtual, as you can see in the code; otherwise it can be really slow.

guolinke replied:
OK, I see. 
Can you change the original ```RawGet``` here : https://github.com/Microsoft/LightGBM/blob/master/src/io/sparse_bin.hpp#L149 to the ```InnerRawGet``` and return ```VAL_T``` type ?

huanzhang12 replied:
Sure. Just to make sure, `InnerRawGet` is just the same as `RawGet` with `VAL_T` return type?

guolinke replied:
yeah

huanzhang12 replied:
So I will merely copy the code of `RawGet()` and make a `InnerRawGet()`?
I guess I can just call `InnerRawGet()` in `RawGet()`? `RawGet()` actually is not used anywhere right now for sparse bins, and it just needs to be there for making the class not pure.
```
template <typename VAL_T>
inline uint32_t SparseBinIterator<VAL_T>::RawGet(data_size_t idx) {
  return InnerRawGet(idx);
}

template <typename VAL_T>
inline VAL_T SparseBinIterator<VAL_T>::InnerRawGet(data_size_t idx) {
  while (cur_pos_ < idx) {
    bin_data_->NextNonzero(&i_delta_, &cur_pos_);
  }
  if (cur_pos_ == idx) {
    return bin_data_->vals_[i_delta_];
  } else {
    return 0;
  }
}
```
Do you think the code above looks okay?

guolinke replied:
yes. However, actually you need a different iterator. This iterator is for one feature not one group.
you can change the related methods in feature_group and dataset to get the correct one.

wxchan commented:
I think we can exclude this submodule, add `--exclude=./compute` at the end of pep8 check.

huanzhang12 commented:
@wxchan Thanks, now the checks have passed.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1207 @@
+    kernel_name = "histogram256";
+    device_bin_size_ = 256;
+    dword_features_ = 4;
+  }
```
dword_features_  are based on the max number of bins? 
If a feature only contains 0/1 values(will use 4bits_bin) but the max_bin set to 255, can this work?

huanzhang12 replied:
Thanks for pointing out this! I think it should work. The code (line 564) checks FeatureNumTotalBin() for each feature, and the variable max_num_bin_ is the maximum number bins. FeatureNumTotalBin() returns the attribute num_total_bin_ of each FeatureGroup. So max_num_bin_ is not the command line parameter max_bin. It depends on the real number of bins used. 

If a feature only contains 0/1 values, then I think max_num_bin_ will be 2. If all features are like this, then max_num_bin_ will be set to 2, and dword_features_ will be set to 8. In this case, all features are using 4bits_bin, and the feature bundles will be created accordingly (line 371). 

If one feature only contains 0/1 values, but some other features have 255 bins, then max_num_bin_ will be set to 255, and dword_features_ will be set to 4. In this case, some features use 4bits_bin and some features use the 8-bit DenseBin. When creating the feature bundles, we have to check if each feature is using a 4bits_bin or not, as you can see around line 406. 

I actually found another small bug during testing this, which is fixed in `15c97b4c`.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1207 @@
+  }
+  if(max_num_bin_ == 17) {
+    Log::Warning("Setting max_bin to 15 is sugguested for best performance");
+  }
```
why ```63``` and ```15``` here? (above is ```16``` and ```64```)

huanzhang12 replied:
When max_bin is set to 64, there can be 65 actual bins created (FeatureNumTotalBin() returns 65, as there is an extra "default bin"), so the GPU kernel with 256 bins has to be used. Similarly, when max_bin is set to 16, there can be 17 actual bins created, and GPU kernel with 64 kernels have to be used. When an user says "max_bin=64" (or "max_bin=16"), I think he/she wants to use the 64-bin (or 16-bin) GPU kernel for best performance. Thus I give a warning in case someone accidentally set max_bin to 64 or 16 (should be 63 or 15).

guolinke commented on src/treelearner/gpu_tree_learner.h:
```diff
@@ -0,0 +1,248 @@
+  int dword_features_;
+  /*! \brief total number of dense feature tuples on GPU.
+   * Each feature tuple is 4-byte (4 features if each feature takes a byte) */
+  int num_dense_feature4_;
```
`num_dense_feature4_` is the number of features with 4_bits_bin?

huanzhang12 replied:
`num_dense_feature4_` is the number of 4-byte feature tuples (or bundles).
Say if we have 30 dense features, and each feature is one byte, then we need to fit the 30 dense features into 8 feature bundles, `num_dense_feature4_` is 8.
If each feature is 4-bit, then `num_dense_feature4_` is 4.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1210 @@
+void GPUTreeLearner::WaitAndGetHistograms(FeatureHistogram* histograms, const std::vector<int8_t>& is_feature_used) {
+  HistType* hist_outputs = (HistType*) host_histogram_outputs_;
+  // when the output is ready, the compuataion is done
+  histograms_wait_obj_.wait();
```
typo: computation

huanzhang12 replied:
Thanks for finding these typos. I will fix all of them in one commit.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1210 @@
+
+void GPUTreeLearner::AllocateGPUMemory() {
+  num_dense_features_ = 0;
+  for (int i = 0; i < num_features_; ++i) {
```
I am not sure about this part. Actually, we build a histogram for a feature_group(a bundle of features), not just for one feature. Refer to : https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L392-L449 .
But it seems you use histogram by per feature (And use RawGet ?) .

huanzhang12 replied:
You're right. When I wrote those code, I used LightGBM v1 as the code base. FeatureGroup seems to be a new addition in LightGBM v2, and honestly I don't think I fully understand it. 

Basically we are trying to replace the line `feature_groups_[group]->bin_data_->ConstructHistogram()` in the function `ConstructHistogram()` of dataset.cpp on GPU. The GPU feature histogram algorithm is not aware of the feature groups, just like the ConstructHistogram() function in `DenseBin` and `Dense4bitsBin` classes. The RawGet() just tries to read the raw data in `DenseBin` or `Dense4bitsBin`, regardless of the feature groups. Then the GPU will do the ConstructHistogram().

I think I probably need your help on fixing this. Could you briefly explain to me how to correctly handle feature groups? I found little comments and explanations on this part of code.

guolinke replied:
A simple explanation: 
1. When building histogram, you can just use the bin_data in a feature_group to build a histogram. However, this histogram actually is the merging of many features (feature_index are continuous). And the histograms of the features in this group can be restored(By calling `Dataset::FixHistogram`) from this merged histogram. Refer to : https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/feature_group.h#L29-L49 
2. when finding best split, you should use the original features histograms.

huanzhang12 replied:
Thanks for the explanation! I like the idea of using feature groups to speed up training sparse features. I will look into the feature group problem further.
BTW, what dataset do you use for testing feature groups during development?

guolinke replied:
new20.binary in this page: https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html

huanzhang12 replied:
I see. That's why my current implementation works even without considering feature groups.

Is it easy to enable feature bundling in master branch? I think I need to get it enabled to fully test my code. 

BTW, is feature bundling enabled only for sparse bins, or it affects dense bins as well? If it is going to be enabled for sparse bins only, it probably does not matter for the time being.

guolinke commented:
As far I can see, This work use GPU to accelerate histogram building. Thus, Can we remove the gpu_tree_learner, and put the construction into Dataset::ConstructHistograms ? By doing this, the parallel version tree learner can benefit from GPU accelerate as well. 
And you can add some necessary functions into ```Dataset``` to do the initialize works.

huanzhang12 commented:
@guolinke Currently it is possible to merge all GPU code into class Dataset, although it possibly makes certain optimization for GPU harder to implement (like asynchronous data transfer). But I think the main issue here is that we probably want to do more on GPU in near future. For example, it is ideal to let GPU do the FindBestThreshold(), and probably entirely avoid copying the histogram back to CPU. If we put the GPU code to class Dataset, implementing these new improvements will be very hard. So it is not ideal in the long run.

I think we can probably let the distributed (parallel) learner somehow inherit from GPUTreeLearner, probably using multiple inheritance? Do you think it is possible?

guolinke commented:
@huanzhang12 
I see. 
Another option is use `SerialTreeLearner` to store the GPU histogram data. 
And the GPU histogram building is done by the ```Dataset``` class. So the ```FindBestThreshold``` can still access the histogram memory in GPU and avoid the copy back to CPU. (Just like the current implementations, the histogram data are stored in the ```SerialTreeLearner```)

Currently, `GpuTreeLearner` has many duplicate codes that exist in the `SerialTreeLearner` . It is less maintainable. 

Also, I think we can just use ```USE_GPU``` to decide use GPU or not. So you can just use macros to rewrite existing methods.

huanzhang12 commented:
@guolinke I agree that `GPUTreeLearner` has too much duplicated code. I have cleaned things up in a039a3a and 35ab97f. Now there is no duplicated code except in function `FindBestThresholds()`. This function deals with GPU directly so it has to be maintained anyways. For best GPU performance, our long term goal is to move most things in this function (like `FixHistogram`, `FindBestThreshold`) to GPU, and there will be just a single GPU kernel call to do everything.

I am still thinking what is the best way to reorganize the GPU related code and how to invoke GPU in parallel learners. The way you proposed sounds okay, but splitting GPU code into two places might also be non-ideal. I will first fix the feature group related issue and then work on this.

huanzhang12 commented:
@guolinke I have adapted my code to the new feature group data structure. Please review the new code. Thanks!

huanzhang12 commented:
@guolinke I propose the following way to make the GPU tree learner available to parallel learners.

First, make `ConstructHistograms` a virtual member function of `TreeLearner`. In `SerialTreeLearner` it will just call the corresponding function in `Dataset`; in `GPUTreeLearner` this function is overridden to construct histograms on GPU. All calls to `ConstructHistograms` in the three parallel learners (feature-based, data-based, voting-based) will be changed to call this member function instead.

Second, make the parallel learners classes as template classes, which are derived from `TREELEARNER_T` instead of `SerialTreeLearner`. For example:
```
template <typename TREELEARNER_T>
class FeatureParallelTreeLearner: public TREELEARNER_T {
public:
   ...
};
```

Third, instantiate different template classes depending if GPU is enabled at runtime. For example:
```
...
  } else if (type == std::string("feature")) {
    if (gpu_is_enabled) {
#ifdef USE_GPU
      return new FeatureParallelTreeLearner<GPUTreeLearner>(tree_config);
#else
      // print some error messages
#endif
    }
    else {
      return new FeatureParallelTreeLearner<SerialTreeLearner>(tree_config);
    }
  } else if (type == std::string("data")) {
...
```
@guolinke How do you like the above approach? If you think it is okay, I can implement this ;)

huanzhang12 commented on src/treelearner/gpu_tree_learner.h:
```diff
@@ -0,0 +1,248 @@
+#ifndef LIGHTGBM_TREELEARNER_GPU_TREE_LEARNER_H_
```
added in 35ab97f

guolinke commented:
@huanzhang12 Thanks.
some questions.
1. Can we just use macros to enable GPU codes ?  Not other tree_learner and setting. If want to run non-GPU version, just compile one without GPU support. 
2. You only can call static function of TREELEARNER_T. So the solution you provide may cannot work. I prefer you create a new ConstructHistograms(only construct histograms), and call it in the derived classes. I think I can do this part. I will update the master branch, and you can merge it.

guolinke commented:
@huanzhang12 
Check the latest code in the master branch. Now the all tree learner share the same code of ConstructHistograms Logic (in SerialTreeLearner::ConstructHistograms).

huanzhang12 commented:
@guolinke Thanks for the comments and code update! I am working on merging to the latest master.

I personally think we should include the capability of enabling/disabling GPU at runtime when GPU support is compiled in, as long as it is not too hard. GPU may not work better than CPU on certain datasets (like a very sparse dataset), so it will be more convenient if we can just set a switch in configuration file to use GPU or not, rather than using two different executable. Especially, when one is using Python/R interface, it is not very straightforward to switch between two shared libraries.

I am not sure why we can only call static functions of TREELEARNER_T. Can you briefly explain it? Thanks!

guolinke commented:
@huanzhang12 
OK. Can we just use a parameters ```use_gpu``` to switch it? 

Check the code in https://github.com/Microsoft/LightGBM/blob/master/src/metric/regression_metric.hpp 
The nested Template class is not a instance, so you cannot call its non-static methods.

guolinke commented:
@huanzhang12 
BTW, I am trying to reduce the sumup cost when hessians are constant: https://github.com/Microsoft/LightGBM/tree/constant-hessian

Is this easy for you to adapt?

huanzhang12 commented:
@guolinke The constant hessian can be adapted, and it will help GPU performance (as the local memory is quite limited on GPU). But it will probably take me some time to fully test the GPU code.

Yes, we can just add a simple switch like `use_gpu`, or probably `device=gpu` (defaulting to `device=cpu`, in future we probably have something like `device=fpga`).

I guess the problem you had in https://github.com/Microsoft/LightGBM/blob/master/src/metric/regression_metric.hpp is probably a different one. It seems I am able to rewrite the parallel learners using the TREELEARNER_T template class without problem. I can't run it for now (no distributed computing environment), but it compiles. You can take a look at the code here: https://github.com/huanzhang12/LightGBM/tree/parallel_gpu_concept If it looks good to you I will try to find some machines to test the code.

guolinke commented:
Got it. I mis-understand it. It can work by this solution.

guolinke commented on src/io/config.cpp:
```diff
@@ -353,6 +358,9 @@ void BoostingConfig::GetTreeLearnerType(const std::unordered_map<std::string, st
       tree_learner_type = "serial";
+    } else if (value == std::string("gpu")) {
+      tree_learner_type = "serial";
+      device_type = "gpu";
```
still need ```tree_learner=gpu``` ?

huanzhang12 replied:
It is not necessary, just another shortcut and intuitive way to enable GPU. If you don't like this I can remove it.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1064 @@
+void GPUTreeLearner::ConstructHistograms(const std::vector<int8_t>& is_feature_used, bool use_subtract) {
+  std::vector<int8_t> is_sparse_feature_used(num_features_, 0);
+  std::vector<int8_t> is_dense_feature_used(num_features_, 0);
+  for (int feature_index = 0; feature_index < num_features_; ++feature_index) {
```
change all loop with ```num_features_``` and ```num_group``` (like https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/serial_tree_learner.cpp#L450-L459 and ) to multi-threading? Or it will be slow when #feature is large.

huanzhang12 replied:
OK

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1064 @@
+  }
+  // converted indices in is_feature_used to feature-group indices
+  std::vector<int8_t> is_feature_group_used(num_feature_groups_, 0);
+  for (int i = 0; i < num_features_; ++i) {
```
multi-threading here? refer to https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp#L437-L448

huanzhang12 replied:
OK

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1064 @@
+  }
+  // construct smaller leaf
+  HistogramBinEntry* ptr_smaller_leaf_hist_data = smaller_leaf_histogram_array_[0].RawData() - 1;
+  bool use_gpu = ConstructGPUHistogramsAsync(is_feature_used,
```
only when ```num_data==0``` will return false? The return name ```use_gpu``` is confusing.

huanzhang12 replied:
There are some cases GPU will not be actually used, like `num_data==0` or all dense features are disabled (I will add this). In these cases `ConstructGPUHistogramsAsync` returns false, indicating that we don't need to call `WaitAndGetHistograms`, otherwise it will deadlock. I can change the variable name and add some comments to the code.

guolinke commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1064 @@
+    leaf = larger_leaf_splits_->LeafIndex();
+    auto larger_best_idx = ArrayArgs<SplitInfo>::ArgMax(larger_best);
+    best_split_per_leaf_[leaf] = larger_best[larger_best_idx];
+  }
```
These duplicate code is only for debug purpose ?

guolinke replied:
If yes, can you just call SerialTreelearner::FindBestThresholds(), and write the additional codes for debugging?

huanzhang12 replied:
Yes, for now we can actually just use `SerialTreelearner::FindBestThresholds()`. But we should keep in mind that later on we also want to do `FixHistograms` and `FindBestThreshold` on GPU as well, without copying the histograms back to GPU.

huanzhang12 commented on src/treelearner/gpu_tree_learner.cpp:
```diff
@@ -0,0 +1,1064 @@
+      for (data_size_t i = 0; i < num_data; ++i) {
+        ordered_gradients[i] = gradients[data_indices[i]];
+      }
+      gradients_future_ = queue_.enqueue_write_buffer_async(device_gradients_, 0, num_data * sizeof(score_t), ptr_pinned_gradients_);
```
`BeforeFindBestSplit` just copies the ordered_grad and ordered_hess for the smaller leaf. For the larger leaf, if subtraction is not possible, ordered_grad and ordered_hess has to be copied before launching the GPU kernel.

guolinke replied:
OK. Why not doing all copy in the GPUTreeLearner::ConstructGPUHistogramsAsync?

huanzhang12 replied:
Because I want to start data copying process as early as possible, hopefully overlap the copying process with other task on CPU and hide the data transfer cost. Currently, in most cases the copying procedure entirely overlaps with `SerialTreeLearner::tBeforeFindBestSplit`. We only wait for these event futures (`indices_future_`, `gradients_future_` and `hessians_future_`) immediately before launching the GPU kernel, and hopefully at that time all data has been transferred to GPU and the overhead for data transfer is close to 0.

huanzhang12 commented:
@guolinke I have added "omp parallel for" to the loops you mentioned, as well as some other places.

I found that for the two loops in `ConstructGPUHistogramsAsync`, the workload of the loop is too small, and it actually harms performance to do multithreading. For `yahoo` and `epsilon`, the entire training can be 20% slower (reproducible). So I added an additional `if` clause for OpenMP, enabling parallelization only when the loop count is large enough.

I also need to add constant hessian optimization to the GPU kernel before merging. But the GPUTreeLearner class will mostly remain unchanged.

guolinke commented:
@huanzhang12 
Thanks, I also fix some other "light" omp loops by using ```if``` .

huanzhang12 commented:
@guolinke I have added constant hessian optimization to the GPU kernels. I changed the interface of TreeLearner a little bit: I also pass the `is_constant_hessian` flag to `Init`, as the `GPUTreeLearner` needs to compile different GPU code based on this flag during initialization and then prepare a lot of stuff before training starts. In function `Train`, if a different value `is_constant_hessian` is passed, it will recompile the GPU kernels accordingly. For other tree learners, it does not matter. Do you think this is okay?

huanzhang12 commented:
@guolinke I am also considering add TravisCI test for GPU, in case that some further updates in other parts break GPU functionality. In the testing environment there is no real GPU, but thanks to the good portability of OpenCL, it is possible to test the GPU code on CPU device after installing a CPU OpenCL environment (like AMD APP SDK or `pocl`). Ideally, all test should pass when `device` is set to both `cpu` and `gpu`. We don't need to change other test options.

What is the best way to globally set a parameter for LightGBM? For example I want to set `device=gpu` and then run all existing testing scripts. I don't want to do heavy modifications to these scripts and add `device=xxx` manually to every tests. In Theano, it takes an environment variable `THEANO_FLAGS` which can easily override a config option. Is there a similar thing in LightGBM?

guolinke commented:
@huanzhang12 I think it didn't have this feature. But it should be easy to add one.

@wxchan Can you add this into python package?

guolinke commented:
@huanzhang12 
I think the test can be done by another PR after finishing this.

huanzhang12 commented:
@guolinke Sure. Currently when I manually do the test there is one error when GPU is enabled (`test_continue_train_multiclass`). I will fix this. After that this PR should be ready ;)

wxchan commented:
@guolinke ok. Do you mean add an argument `is_constant_hessian` to sklearn `__init__`?

guolinke commented:
@wxchan No, I mean add something like ```THEANO_FLAGS```(environment variable) to choose cpu/gpu device when call lightgbm.dll . I think it can be solved by detect the environment variable, and change the related parameters passed to lightgbm.dll .

huanzhang12 commented:
@guolinke The variable `filter_cnt` passed to function `BinMapper::FindBin` seems to be calculated in different ways at the two places in `dataset_io.cpp`:

At https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset_loader.cpp#L491-L492
```
  const data_size_t filter_cnt = static_cast<data_size_t>(
    static_cast<double>(io_config_.min_data_in_leaf * total_sample_size) / num_data);
```
Where `total_sample_size` seems to be the total number of examples;

But at https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset_loader.cpp#L709-L710
```
  const data_size_t filter_cnt = static_cast<data_size_t>(
    static_cast<double>(io_config_.min_data_in_leaf* sample_values.size()) / dataset->num_data_);
```
`sample_values.size()` is the number of features.

I found this issue when debugging the failure test case. I guess one of them is a typo?

guolinke commented:
oh, It is typo. I will fix it..

huanzhang12 commented:
@guolinke I have fixed the test failure case. Also I added travisCI tests for GPU leaner. You can make a final review and finish this PR now.

guolinke commented:
@huanzhang12 
It seems XGBoost just support the histogram algorithm on GPU.
Do you have chance to compare it ?

huanzhang12 commented:
@guolinke I haven't got a chance to carefully compare with it yet.
I briefly take a look at the code, and the good thing is that in XGBoost everything is running on GPU (histogram, split, update, etc), so it could be faster than LightGBM (only doing histogram on GPU) when the dataset is small. However I am not sure if the GPU tree learner in XGBoost is exactly the same as the CPU one, because it basically re-implements everything.

The histogram building procedure on XGBoost is not as simple as the one in LightGBM and needs more bookkeeping, so I guess LightGBM could outperform it on large dataset.

tt83 commented:
@huanzhang12 Is it possible to parallelize prediction on GPU?

huanzhang12 commented:
@tt83 Yes I think it is possible. In fact, we can generate OpenCL code like https://github.com/Microsoft/LightGBM/pull/469 for GPU, and run it on GPU for prediction.
